Add example activity for P24/Bancontact/Giropay/EPS

smaskell-stripe opened the pull request:
## Summary
<!-- Simple summary of what was changed. -->
![simple_payment_compound_drawable](https://user-images.githubusercontent.com/47332718/80138972-90233a00-855a-11ea-8613-a00226595243.gif)




## Motivation
<!-- Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue. -->

## Testing
<!-- How was the code tested? Be as specific as possible. -->

smaskell-stripe commented on example/src/main/java/com/stripe/example/activity/SimpleRedirectPaymentMethodActivity.kt:
```diff
@@ -0,0 +1,205 @@
+            )
+        }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
I should probably handle saved instance state here

mshafrir-stripe replied:
Totally optional. If you want, you can try creating a `ViewModel` to handle that for you.

mshafrir-stripe commented on example/res/layout/simple_redirect_activity.xml:
```diff
@@ -0,0 +1,82 @@
+            android:layout_gravity="center_horizontal"
+            android:layout_marginTop="20dp"
+            android:enabled="true"
+            android:text="Submit" />
```
Let's change the text label to something like "Pay now"

And we can also add support for Setup Intents with a "Save for later" button

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/SimpleRedirectPaymentMethodActivity.kt:
```diff
@@ -0,0 +1,205 @@
+    private val keyboardController: KeyboardController by lazy {
+        KeyboardController(this)
+    }
+    private val compositeSubscription = CompositeDisposable()
+    private val backendApi: BackendApi by lazy {
+        BackendApiFactory(this).create()
+    }
```
We do this in a lot of examples, I wonder if we should create a `StripeExampleActivity` superclass. Doesn't have to be in this PR.

smaskell-stripe replied:
Yeah, I was wondering the same thing

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/LauncherActivity.kt:
```diff
@@ -65,7 +65,9 @@ class LauncherActivity : AppCompatActivity() {
             Item(activity.getString(R.string.card_brands),
-                CardBrandsActivity::class.java)
+                CardBrandsActivity::class.java),
+            Item("Simple Redirect PaymentMethod Confirmation Example",
+                SimpleRedirectPaymentMethodActivity::class.java)
```
Maybe "Simple PaymentMethod Confirmation Example" is enough?

smaskell-stripe replied:
Sure. Names are hard

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/SimpleRedirectPaymentMethodActivity.kt:
```diff
@@ -0,0 +1,205 @@
+        paymentIntentClientSecret: String,
+        params: PaymentMethodCreateParams
+    ) {
+        viewBinding.status.append("\n\nStarting payment authentication")
```
Let's change this to "Starting PaymentIntent confirmation"

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/SimpleRedirectPaymentMethodActivity.kt:
```diff
@@ -0,0 +1,205 @@
+    private fun onConfirmSuccess(result: PaymentIntentResult) {
+        val paymentIntent = result.intent
+        viewBinding.status.append("\n\n" +
+            "Auth outcome: ${result.outcome}\n\n" +
```
"PaymentIntent confirmation outcome: ..."

You can also change this to a multiline string literal

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/SimpleRedirectPaymentMethodActivity.kt:
```diff
@@ -0,0 +1,205 @@
+            backendApi
+                .createPaymentIntent(
+                    mutableMapOf(
+                        "amount" to 1000,
```
you can remove `amount` - it's calculated dynamically on the backend

smaskell-stripe commented:
@mshafrir-stripe I think I addressed all feedback. Also added icons to the dropdown as we discussed.

However, the setup intents don't work right now. Possibly need to change something on the backend to allow them for these payment types?
I get an error saying `The payment method type provided: <type> is invalid`

smaskell-stripe commented:
Removed the support for SetupIntents since they do not exist for these payment methods

mshafrir-stripe commented on example/res/layout/dropdown_menu_popup_item.xml:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Nit: could you use simplify this to a `TextView` with a compound drawable?

mshafrir-stripe commented on example/res/layout/simple_payment_method_activity.xml:
```diff
@@ -0,0 +1,82 @@
+                android:id="@+id/payment_method"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:editable="false"
```
this attribute appears to be deprecated, can you use `inputType` instead?

smaskell-stripe replied:
unfortunately, that doesn't have the same effect (`inputType="none"` *should* disable editing but it doesn't on this widget for some reason). I kind of regret even trying to use this widget altogether as it's all kinds of wrong. Really this should be a spinner but there is no material spinner - https://medium.com/@rmirabelle/there-is-no-material-design-spinner-for-android-3261b7c77da8

mshafrir-stripe commented on example/src/main/java/com/stripe/example/activity/SimplePaymentMethodConfirmationActivity.kt:
```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putString(STATE_EMAIL, viewBinding.email.text.toString())
+        outState.putString(STATE_NAME, viewBinding.name.text.toString())
+        outState.putString(STATE_DROPDOWN_ITEM, viewBinding.paymentMethod.text.toString())
+    }
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
+        super.onRestoreInstanceState(savedInstanceState)
+        savedInstanceState.getString(STATE_EMAIL)?.let(viewBinding.email::setText)
+        savedInstanceState.getString(STATE_NAME)?.let(viewBinding.name::setText)
+        savedInstanceState.getString(STATE_DROPDOWN_ITEM)?.let {
+            viewBinding.paymentMethod.setText(it)
+            onDropdownItemSelected()
+        }
+    }
```
this is totally fine, but you could also take advantage of the `ViewModel` to preserve state across rotation

smaskell-stripe replied:
Yeah, I looked into doing that but it didn't really seem any better than just doing this because the data isn't coming from the view model (it's coming from the user). I'd have to add a text changed listener to each view and store it on the view model every time something changes. Or am I missing something here?

mshafrir-stripe replied:
You could add the vars on the ViewModel and populate when `onSaveInstanceState()` is called, then read and update view in `onCreate()`. There's probably no good reason to do this here though.

smaskell-stripe commented:
Okay. One more pass? 🙏
